### ISSUE TITLE
Invalidate token when the private code of a project is changed?

### ISSUE THREAD

**Issue by zorun** [spiral-project/ihatemoney#780]
Currently, the token associated to the project is generated based on the global secret key and the project ID, and is independent of the private code.

It means that the following is possible:

- a project is created with private code A
- somebody gets access to the token (either through the API, or simply by receiving an invitation email because it contains a link with the token)
- the private code of the project is changed to B (e.g. to prevent old roommates from accessing the project)
- the old token is still valid and still allows full access to the project: just clicking the link of an old invitation email is enough to get access

It doesn't sound very good to me, but maybe I'm missing something?

It would be simple to fix: just add the private code as additional input to the token generation.

**Comment by Glandos** [comment]
Yes, it seems a bit logical that a new private code invalidate old tokens. We could even imagine to use the last code update timestamp (as available with continuum) to be sure that setting a new code and going back to an already used code doesn't reactivate tokens.

**Comment by zorun** [comment]
When going back to a previous private code, then it would make sense that the associated token works again, no?

Basically, the token would be a kind of secure hash of project ID + private code, so it would have exactly the same sensitivity as the private code.

**Comment by Glandos** [comment]
[**@zorun**](https://github.com/zorun) commented on [7 juil. 2021 à 22:50 UTC+2](https://github.com/spiral-project/ihatemoney/issues/780#issuecomment-875923769 "2021-07-07T20:50:55Z - Replied by Github Reply Comments"):
> When going back to a previous private code, then it would make sense that the associated token works again, no?

I imagined the use case where you shared a token without sharing the private code to someone, and all others know the private code. Now you just want to invalidate the token. So you'll just have to change private code twice.

> Basically, the token would be a kind of secure hash of project ID + private code, so it would have exactly the same sensitivity as the private code.

In fact, you're right. I guess that once you have a token, you can change the private code anyway, thus invalidating others. So if you have a token once, you're part of the project.

Let's keep it simple: token generated from project ID + private code.

An extra security measure could request the **old** code before changing for a **new** one.

**Comment by zorun** [comment]
Well, sorry @Glandos , I hadn't thought about this egg-and-chicken problem: to validate the token, we need to know the project ID (so that we can fetch the private code), but we don't know the project ID until we have decoded the token...

Maybe we can take another approach with a counter:

- store a counter in the database for each project
- when the private code of a project is changed, increase the counter
- the token payload would be `{"project_id": "foo", "counter": 42}`
- the secret key used to sign a token would be just `SECRET_KEY`, as it is currently
- when verifying a token, it would take three steps:
 1) (unchanged) verify crypto signature
 2) (unchanged) fetch project ID from the payload, verify it exists in the database
 3) (new) fetch counter from the payload, verify it's equal to the counter in the database for this project

This achieves the goal: each time the private code is changed, the existing token is invalidated.

It can be backwards compatible: if a token from a previous ihatemoney version is used (without counter), just assume that `counter == 0`. This should make sure that existing tokens still work (only until somebody changes the private code of course).
